test/com/amplitude/api/AmplitudeTest.java

```diff
@@ -53,1073 +31,116 @@ public void tearDown() throws Exception {
-        assertTrue(compareJSONObjects(userProperties, expected));
-    }
-
-    @Test
-    public void testReloadDeviceIdFromDatabase() {
-        String deviceId = "test_device_id";
-        ShadowLooper looper = Shadows.shadowOf(amplitude.logThread.getLooper());
-
-        assertNull(amplitude.getDeviceId());
-        DatabaseHelper.getDatabaseHelper(context).insertOrReplaceKeyValue(
-                AmplitudeClient.DEVICE_ID_KEY,
-                deviceId
-        );
-        looper.getScheduler().advanceToLastPostedRunnable();
-        assertEquals(deviceId, amplitude.getDeviceId());
-    }
-
-    @Test
-    public void testDoesNotUpgradeDeviceIdFromSharedPrefsToDatabase() {
-        assertNull(amplitude.getDeviceId());
-        ShadowLooper looper = Shadows.shadowOf(amplitude.logThread.getLooper());
-
-        // initializeDeviceId no longer fetches from SharedPrefs, will get advertising ID instead
-        String targetName = Constants.PACKAGE_NAME + "." + context.getPackageName();
-        SharedPreferences prefs = context.getSharedPreferences(targetName, Context.MODE_PRIVATE);
-        prefs.edit().putString(Constants.PREFKEY_DEVICE_ID, "test_device_id").commit();
-
-        looper.getScheduler().advanceToLastPostedRunnable();
-        String deviceId = amplitude.getDeviceId();
-        assertTrue(deviceId.endsWith("R"));
+    public void testGetInstance() {
+        AmplitudeClient a = Amplitude.getInstance();
+        AmplitudeClient b = Amplitude.getInstance("");
+        AmplitudeClient c = Amplitude.getInstance(null);
+        AmplitudeClient d = Amplitude.getInstance(Constants.DEFAULT_INSTANCE);
+        AmplitudeClient e = Amplitude.getInstance("app1");
+        AmplitudeClient f = Amplitude.getInstance("app2");
+
+        assertSame(a, b);
+        assertSame(b, c);
+        assertSame(c, d);
+        assertSame(d, Amplitude.getInstance());
+        assertNotSame(d, e);
+        assertSame(e, Amplitude.getInstance("app1"));
+        assertNotSame(e, f);
+        assertSame(f, Amplitude.getInstance("app2"));
+
+        assertTrue(Amplitude.instances.size() == 3);
+        assertTrue(Amplitude.instances.containsKey(Constants.DEFAULT_INSTANCE));
+        assertTrue(Amplitude.instances.containsKey("app1"));
+        assertTrue(Amplitude.instances.containsKey("app2"));
+    }
+
+    @Test
+    public void testSeparateInstancesLogEventsSeparately() {
+        Amplitude.instances.clear();
+        DatabaseHelper.instances.clear();
+
+        String newInstance1 = "newApp1";
+        String newApiKey1 = "1234567890";
+        String newInstance2 = "newApp2";
+        String newApiKey2 = "0987654321";
+        File oldDbFile = context.getDatabasePath(Constants.DATABASE_NAME);
+        File newDbFile1 = context.getDatabasePath(Constants.DATABASE_NAME + "_" + newInstance1);
+        File newDbFile2 = context.getDatabasePath(Constants.DATABASE_NAME + "_" + newInstance2);
```

Comment:
This can probably go either way, but it might be cleaner to have a ```DatabaseHelper``` reference expose an existence check instead of having manually checking the file?